app/metrics.py

```diff
@@ -44,24 +45,27 @@ def get_previous_metrics(w3, pool, oracle, genesis_time, epochs_per_frame, from_
     latest_block = w3.eth.getBlock('latest')
     from_block = max(from_block, int((latest_block['timestamp']-genesis_time)/SECONDS_PER_ETH1_BLOCK))
     step = 10000
-    # Fetch and parse 'Completed' events from the contract.
+    # Try to fetch and parse last 'Completed' event from the contract.
     for end in range(latest_block['number'], from_block, -step):
         start = max(end - step + 1, from_block)
         events = oracle.events.Completed.getLogs(fromBlock=start, toBlock=end)
         if events:
             event = events[-1]
             result.epoch = event['args']['epochId']
-            block = w3.eth.getBlock(event['blockHash'])
-            result.timestamp = block['timestamp']
-            return result
+            break
 
-    # If the list of events is empty, we consider it's the first run
-    result.epoch = 0
-    result.timestamp = latest_block['timestamp']
+    # If the epoch has been assigned from the last event (not the first run)
+    if result.epoch:
+        result.timestamp = get_timestamp_by_epoch(beacon_spec, result.epoch)
+    else:
+        # If it's the first run, we set timestamp to genesis time
```

Comment:
Don't we need to assign something to result's epoch?

Reply:
It's assigned in the class https://github.com/lidofinance/lido-oracle/blob/cf0622714752deaf88866d4ea92a114c4e55cf64/app/metrics.py#L14 . The default value of 0 is a pretty good marker for future use on consumer side (to check for corner-cases similar to https://github.com/lidofinance/lido-oracle/blob/cf0622714752deaf88866d4ea92a114c4e55cf64/app/metrics.py#L122 and https://github.com/lidofinance/lido-oracle/blob/cf0622714752deaf88866d4ea92a114c4e55cf64/app/metrics.py#L126)